networkx/algorithms/bipartite/projection.py

```diff
@@ -185,33 +185,69 @@ def weighted_projected_graph(B, nodes, ratio=False):
         Networks". In Carrington, P. and Scott, J. (eds) The Sage Handbook
         of Social Network Analysis. Sage Publications.
     """
-    if B.is_directed():
-        pred = B.pred
-        G = nx.DiGraph()
+
+    if type(nodes) != list:
+        top_nodes = []
+        for node in B.nodes():
+            if type(node) != type(nodes):
+                top_nodes.append(node)
+            elif node not in nodes:
+                top_nodes.append(node)
     else:
-        pred = B.adj
-        G = nx.Graph()
-    G.graph.update(B.graph)
-    G.add_nodes_from((n, B.nodes[n]) for n in nodes)
-    n_top = len(B) - len(nodes)
+        top_nodes = [node for node in B.nodes() if node not in nodes]
 
-    if n_top < 1:
+    if len(top_nodes) + len(nodes) != B.number_of_nodes():
         raise NetworkXAlgorithmError(
             f"the size of the nodes to project onto ({len(nodes)}) is >= the graph size ({len(B)}).\n"
             "They are either not a valid bipartite partition or contain duplicates"
         )
 
-    for u in nodes:
-        unbrs = set(B[u])
-        nbrs2 = {n for nbr in unbrs for n in B[nbr]} - {u}
-        for v in nbrs2:
-            vnbrs = set(pred[v])
-            common = unbrs & vnbrs
-            if not ratio:
-                weight = len(common)
-            else:
-                weight = len(common) / n_top
-            G.add_edge(u, v, weight=weight)
+    if not B.is_directed():
+        # Get the biadjacency matrix.
+        adj = nx.bipartite.biadjacency_matrix(
+            B, row_order=nodes, column_order=top_nodes
+        )
+        G = nx.Graph()
+        G.add_nodes_from(nodes)
+        node_labels = {i: nodes[i] for i in range(len(nodes))}
```

Comment:
Maybe just `node_labels = dict(enumerate(nodes))`? 

And while we're mapping indexes to nodes, we could map nodes to indexes and use that dict as a lookup tool instead of `<list>.index()` as well as replacing `node in top_nodes` by checking `node not in nodes_indexes`, and replacing `len(top_nodes)`.

Then, another speed-up might involve each place you use `for i in range(len(<list>))`. You should probably use `enumerate` to get both the index and the value so they are both available in that loop.

Reply:
Made the change to `node_labels = dict(enumerate(nodes))`. Also, used it for the lookup.